derp/derp_server.go

```diff
@@ -1398,3 +1418,107 @@ func writePublicKey(bw *bufio.Writer, key *key.Public) error {
 	}
 	return nil
 }
+
+type ipBytesSent struct {
+	ip        netaddr.IP
+	bytesSent int64
+	lastSeen  time.Time
+
+	index int
+}
+
+func (i ipBytesSent) value(maxTime time.Duration) float64 {
+	decay := math.Max(0, float64(maxTime-time.Since(i.lastSeen))/float64(maxTime))
+	return float64(i.bytesSent) * decay
+}
+
+type ipBytesSentPriorityQueue struct {
```

Comment:
I still think this is going to be too slow to be in the data path.

You don't want to do a mutex or a map lookup.

I'd just increment a local value on `sclient`, and only add/remove it from a leaderboard set if the value passes a certain value

Reply:
hm makes sense, that's why I put it in a goroutine so that it wouldn't block the main data path but it's pretty costly still